--- 6.72f.j	2011-12-28 15:55:10.000000000 +0100
+++ 6.73b.j	2011-12-28 15:53:55.000000000 +0100
@@ -453,59 +453,59 @@
 sound gg_snd_WallOfReplica=null
 sound gg_snd_WhickedSick=null
 sound gg_snd_GoblinSapperYesAttack4=null
 sound gg_snd_ShadowStrikeBirth1=null
 sound gg_snd_TheBlackArrow2=null
 sound gg_snd_CorrosiveBreathMissileLaunch1=null
 sound gg_snd_Death_Sentinel=null
 sound gg_snd_Death_Scourge=null
 sound gg_snd_ReviveHuman=null
 sound gg_snd_SecretFound=null
 sound gg_snd_UltraKill=null
 sound gg_snd_Rampage=null
 sound gg_snd_AltarOfEldersWhat1=null
 sound gg_snd_MysticSnakeCast=null
 sound gg_snd_MysticSnakeHit=null
-sound gg_snd_SotS2=null
@@ -2689,30 +2718,37 @@
 constant integer ancientjangoo_boost_buffid=1110459480
 constant real ancientjangoo_duration=6
 constant integer abilityid_smokeofdeceit=1093808182
 constant string smokeofdeceit_fx="war3mapImported\\SmokeOfDeceit.mdx"
 constant integer smokeofdeceit_invis=1093808204
 constant integer smokeofdeceit_ms=1093808211
 constant integer smokeofdeceit_buffid=1110459698
 constant integer smokeofdeceit_phasebuff=1110456665
 unit smokeofdeceit_caster
 player smokeofdeceit_player
 constant integer abilityid_veilofdiscord=1093810244
 constant integer veilofdiscord=1093810246
 constant integer veilofdiscord_buffid=1110459729
 constant real veilofdiscord_duration=15
 constant string veilofdiscord_fx="war3mapImported\\DarkLightningNova.mdx"
+constant integer abilityid_cripplingstaff=1093813569
+constant integer cripplingstaff_miss=1093814861
+constant integer cripplingstaff_miss_buffid=1110459989
+constant integer cripplingstaff_slow=1093813571
+constant integer cripplingstaff_slow_buffid=1110459957

+constant integer abilityid_tranquilboots=1093815089
 constant integer abilityid_corruptor=1093753420
 constant integer abilityid_corruptor_ranged=1093753421
 constant integer unitid_corruptor=1848659021
 constant integer unitid_corruptor_ranged=1848659022
 constant integer abilityid_medallion_of_courage=1093753417
 constant integer medallion_armor=1093753418
 constant string medallion_fx="Abilities\\Spells\\Undead\\Cripple\\CrippleTarget.mdl"
 boolean aegis_meepo=false
 constant integer abilityid_armletofmordiggianoff=1093686327
 constant integer abilityid_armletofmordiggianon=1093686328
 boolean throne_damage_10=false
 boolean throne_damage_25=false
 boolean throne_damage_50=false
 boolean throne_damage_75=false
 boolean tree_damage_10=false
@@ -2964,31 +3001,31 @@
 image CD_Image_14
 image CD_Image_15
 image CD_Image_16
 constant integer CD_PHASE_BAN_1=1
 constant integer CD_PHASE_BAN_2=2
 constant integer CD_PHASE_BAN_3=3
 constant integer CD_PHASE_BAN_4=4
 constant integer CD_PHASE_DRAFT_1=5
 constant integer CD_PHASE_DRAFT_2=6
 constant integer CD_PHASE_DRAFT_3=7
 constant integer CD_PHASE_DRAFT_4=8
 constant integer CD_PHASE_DRAFT_5=9
 constant integer CD_PHASE_DRAFT_6=10
 constant integer CD_Shadow=1747990089
 constant integer CD_unitid_selecthero=1747990094
-constant integer CD_COUNT=24
+constant integer CD_COUNT=28
 constant real CD_RADIUS=450
 constant real CD_HEROSPACE=75
 constant real CD_HERODISTANCE=140
 constant real CD_HEROBANSPACE=80
 constant real CD_HEROBANDISTANCE=210
 constant integer CD_TOTALTIME=150
 constant integer MSG_CD_BAN1=1848657970
 constant integer MSG_CD_PICK1=1848657971
 constant integer MSG_CD_PICK2=1848657972
 constant integer MSG_CD_ACTION_1=1848657973
 constant integer MSG_CD_ACTION_2=1848657974
 constant integer MSG_CD_ACTION_3=1848657975
 constant integer MSG_CD_ACTION_4=1848657976
 constant integer MSG_CD_FINISHED=1848657977
 boolean udg_ModeSD=false
@@ -3091,30 +3128,31 @@
 image array CP_Image_b
 constant integer abilityid_torrent=1093743414
 constant integer torrent_toss=1093743447
 constant integer torrent_slow=1093742924
 constant integer abilityid_watersword=1093743444
 constant integer watersword_damage=1093742927
 constant integer watersword_fx=1093742930
 constant integer watersword_cleave1=1093743671
 constant integer watersword_cleave2=1093743672
 constant integer watersword_cleave3=1093742933
 constant integer watersword_cleave4=1093743673
 constant integer abilityid_xmarksthespot=1093742926
 constant integer abilityid_xmarksthespot_helper=1093743428
 constant integer abilityid_ghostship=1093742923
 constant integer unitid_ghostship=1747990101
+constant integer unitid_ghostship_alt=1747993674
 constant integer unitid_ghostship2=1747990102
 constant integer ghostship_slow=1093742924
 constant integer ghostship_stun=1093743448
 constant integer rum_bonus=1093743158
 constant integer buffid_rum_ms=1110456661
 group ghostship_alreadyhit
 unit ghostship_source
 integer ghostship_level
 real ghostship_angle
 constant integer abilityid_unstableconcoction=1093750345
 constant integer abilityid_unstableconcoction_helper=1093750344
 constant integer unitid_unstableconcoction=1747993156
 constant real unstableconcoction_maxcharge=5.0
 constant real unstableconcoction_buff=2.0
 unit unstableconcoction_source
@@ -3179,49 +3217,50 @@
 constant integer abilityid_hookshotstun=1093687865
 constant integer unitid_hookshot=1966092374
 constant integer unitid_hookshothead=1966092373
 constant integer abilityid_batteryassault=1093687860
 constant integer abilityid_batteryassaultstun=1093687863
 constant string fx_batteryassault="Abilities\\Spells\\Human\\FlakCannons\\FlakTarget.mdl"
 constant integer unitid_batteryassault=1747988570
 constant integer abilityid_freezingfieldcast=1093677912
 constant integer abilityid_freezingfield=1093677906
 constant integer abilityid_freezingfieldupgrade=1093681494
 constant integer abilityid_freezingfield_slow=1093742676
 unit freezingfield_source
 integer freezingfield_damage
 constant integer abilityid_crystalnova=1093748025
 constant integer crystalnova_nuke=1093677380
-constant integer unitid_raijin=1211117651
 constant integer abilityid_staticremnant=1093743696
 constant integer unitid_staticremnant=1747990342
 unit staticremnant_source
 real staticremnant_damage
 constant integer abilityid_overload=1093685591
 constant integer abilityid_overloadslow=1093686069
 unit overload_source
 real overload_damage
 constant integer abilityid_electricvortex=1093743698
 constant integer electricvortex_shackle=1093743697
 constant integer unitid_electricvortex=1966092376
 constant integer electricvortex_slow=1093743699
 constant integer abilityid_balllightning=1093743695
 integer balllightning_cache
 integer balllightning_level
 unit balllightning_source
 real balllightning_traveled
 constant integer abilityid_dragonstail=1093681490
+constant integer abilityid_dragonstail_learn=1093812553
+constant integer abilityid_dragonform=1093677895
 constant integer abilityid_fissure=1093686091
 constant integer abilityid_fissurestun=1093686090
 constant integer abilityid_echoslam=1093682248
 constant integer abilityid_echoslam_upgrade=1093750594
 constant integer abilityid_echoslam_fok=1093682253
 constant integer abilityid_echoslam_fok_half=1093748784
 unit array echoslam_caster
 unit array echoslam_caster_half
 constant integer abilityid_naturesattendants=1093677378
 constant integer abilityid_impetus=1093682265
 constant integer abilityid_impetus_upgrade=1093752642
 constant integer buffid_impetus=1110455125
 constant integer abilityid_enchant=1093682264
 constant integer abilityid_malefice=1093683511
 constant integer abilityid_maleficestun=1093677377
@@ -3472,31 +3511,31 @@
 constant integer shrapnel_slow=1093744966
 constant string fx_shrapnel="Abilities\\Weapons\\FlyingMachine\\FlyingMachineImpact.mdl"
 constant integer unitid_shrapnel_fx=1747990593
 unit shrapnel_source
 unit shrapnel_caster
 real shrapnel_damage
 constant integer abilityid_assassinate=1093678160
 constant integer abilityid_assassinateupgrade=1093747029
 constant integer unitid_smokescreen=1697656883
 constant integer abilityid_smokescreen=1093685831
 constant integer abilityid_smokescreensent=1093682487
 constant integer abilityid_smokescreenscourge=1093677369
 constant integer abilityid_smokescreenstun=1093682486
 constant integer abilityid_blinkstrikesa=1093684025
 constant integer abilityid_backstab=1093682266
-constant integer abilityid_warcry=1093746997
+constant integer abilityid_warcry=1093814611
 constant integer warcry_abilityid_1=1093748278
 constant integer warcry_abilityid_2=1093748279
 constant integer warcry_abilityid_3=1093748275
 constant integer warcry_abilityid_4=1093748277
 constant string warcry_fx="Abilities\\Spells\\NightElf\\BattleRoar\\RoarCaster.mdl"
 integer warcry_abilityid
 constant integer abilityid_godstrength=1093752648
 constant integer abilityid_stormbolt=1093744944
 constant integer unitid_stormbolt=1747990353
 constant integer stormbolt_stun=1093685850
 constant string stormbolt_fx="Abilities\\Spells\\Human\\StormBolt\\StormBoltMissile.mdl"
 constant string stormbolt_fx_attach="origin"
 unit stormbolt_caster
 constant integer abilityid_rabid_1=1093748038
 constant integer abilityid_rabid_2=1093748037
@@ -3548,30 +3587,33 @@
 integer ancestralcharge_attackbonus
 integer ancestralcharge_movementbonus
 unit array ancestralcharge_image
 boolean ancestralcharge_active=false
 boolean echostomp_ignore=false
 constant integer abilityid_suicidesound=1093678658
 trigger remotemine_deathtrigger
 integer remotemine_unitid
 player remotemine_player
 integer minescommand_count
 constant integer abilityid_marchofthemachines=1093681745
 constant integer abilityid_toss=1093681754
 unit toss_source
 real toss_damage
 unit toss_target
+constant integer grow_aghanim_treefix=1093815105
+constant integer grow_aghanim_treefix_to=1093815106
+constant integer grow_aghanim_treefix_from=1093815107
 constant integer abilityid_warclub_fake=1093752392
 constant integer abilityid_warclub_real=1093752391
 constant integer warclub_cleave=1093752625
 constant integer abilityid_grow_bonus1=1093742658
 constant integer abilityid_grow_bonus2=1093742657
 constant integer abilityid_grow_bonus3=1093687891
 constant integer abilityid_grow=1093682009
 constant integer abilityid_craggyexterior=1093744977
 constant integer craggyexterior_stun=1093681750
 constant integer avalanche_stun=1093742666
 constant integer avalanche_visual=1093742669
 constant integer abilityid_avalanche=1093684300
 unit avalanche_source
 real avalanche_damage
 constant integer abilityid_rampage=1093748042
@@ -3658,36 +3700,46 @@
 constant integer abilityid_berserkerscall=1093683510
 unit berserkerscall_unit
 constant integer abilityid_counterhelix=1093681974
 constant integer buffid_counterhelix=1110455120
 constant integer abilityid_cullingblade=1093682482
 constant integer abilityid_cullingblade_upgrade=1093750098
+constant integer cullingblade_haste=1093814595
+constant integer cullingblade_buffid=1110459990
 constant integer abilityid_nightmare=1093678169
 constant integer buffid_nightmare=1110454854
 constant integer abilityid_fiendsgrip=1093677649
 constant integer abilityid_fiendsgripupgrade=1093747769
 constant integer abilityid_enfeeble=1093678166
 constant integer enfeeble_1=1093808689
 constant integer enfeeble_2=1093808691
 constant integer enfeeble_3=1093808692
 constant integer enfeeble_4=1093808690
 constant integer abilityid_stickynapalm=1093748044
 integer array stickynapalm_slow_1
 integer array stickynapalm_slow_2
 integer array stickynapalm_slow_3
 integer array stickynapalm_slow_4
 constant integer buffid_stickynapalm=1110458963
@@ -3716,30 +3768,32 @@
 constant integer abilityid_bloodbath=1093684293
 constant string bloodbath_fx="Objects\\Spawnmodels\\Human\\HumanBlood\\HeroBloodElfBlood.mdl"
 unit bloodbath_target
 constant integer abilityid_rupture=1093684296
 constant integer abilityid_thirst=1093683512
 constant integer thirst_ms_1=1093808432
 constant integer thirst_ms_2=1093808433
 constant integer thirst_ms_3=1093808434
 constant integer thirst_ms_4=1093808435
 integer array thirst_ms
 constant integer buffid_thirst_1=1110455634
 constant integer buffid_thirst_2=1110459701
 constant integer buffid_thirst_3=1110459700
 constant integer buffid_thirst_4=1110459702
 constant integer thirst_fx=1093808436
+constant integer thirst_reveal=1093812552
+constant integer thirst_reveal_buffid=1110459737
 group thirst_group
 group thirst_tempgroup
 unit thirst_source
 integer thirst_level
 constant integer abilityid_bloodrage=1093682499
 constant integer abilityid_insatiablehunger=1093687121
 constant integer abilityid_spawnspiderlings=1093681736
 constant integer buffid_spawnspiderlings=1110459703
 constant integer abilityid_blinkstrike_ck=1093685847
 integer unitid_blinkstrike_ck_1=1747990090
 real blinkstrike_ck_x
 real blinkstrike_ck_y
 unit blinkstrike_ck_source
 unit blinkstrike_ck_target
 integer blinkstrike_debug=1
@@ -4143,30 +4197,34 @@
 constant integer zeal_regen_4=1093809976
 constant integer abilityid_sunder=1093678929
 constant integer abilityid_sundercaster=1093678930
 constant integer abilityid_reflection=1093809985
 constant integer reflection_slow_1=1093809752
 constant integer reflection_slow_2=1093809750
 constant integer reflection_slow_3=1093809753
 constant integer reflection_slow_4=1093809751
 constant integer reflection_slow_buffid=1110459722
 constant integer reflection_image=1093809227
 constant integer reflection_image_buffid=1110459726
 constant string reflection_fx="war3mapImported\\BlackChakraExplosion.mdx"
 constant integer abilityid_metamorph=1093751369
 constant integer metamorph_image=1093679185
 constant integer metamorph_buffid=1110459721
+constant integer abilityid_soulsteal=1093750856
+constant integer abilityid_soulsteal_dummy=1093678156
+constant integer abilityid_soulsteal_secondary=1093751361
+constant integer unitid_soulsteal=1966092615
 constant integer abilityid_krakenshell=1093678149
 constant real krakenshell_maxtime=6
 constant real krakenshell_maxdamage=600
 constant integer abilityid_ravage=1093810505
 constant integer abilityid_anchorsmash=1093808694
 constant integer anchorsmash_attack=1093808696
 constant integer anchorsmash_attack_buffid=1110459704
 constant string anchorsmash_fx="Abilities\\Spells\\Other\\CrushingWave\\CrushingWaveDamage.mdl"
 unit anchorsmash_source
 integer anchorsmash_level
 unit shadowwave_source
 constant integer abilityid_weave=1093742673
 constant integer abilityid_weaveupgrade=1093747778
 constant integer WEAVE_POSITIVE=1
 constant integer WEAVE_NEGATIVE=2
@@ -4224,48 +4282,58 @@
 constant integer gravekeeperscloak_level4_4=1093749826
 constant integer gravekeeperscloak_level4_5=1093749825
 constant integer gravekeeperscloak_level4_6=1093749814
 constant integer gravekeeperscloak_level4_7=1093749832
 constant integer gravekeeperscloakenemy_1=1093750067
 constant integer gravekeeperscloakenemy_2=1093750068
 constant integer gravekeeperscloakenemy_3=1093750083
 constant integer gravekeeperscloakenemy_4=1093750071
 constant integer gravekeeperscloakenemy_5=1093750072
 constant integer gravekeeperscloakenemy_6=1093750073
 constant integer gravekeeperscloakenemy_7=1093750081
 constant integer gravekeeperscloakenemy_8=1093750069
 constant integer gravekeeperscloakenemy_9=1093750082
 constant integer gravekeeperscloakenemy_10=1093750070
 constant integer abilityid_gargoyle=1093750341
+constant integer abilityid_gargoyle_upgrade=1093814599
 constant integer unitid_gargoyle_a_1=1966092596
 constant integer unitid_gargoyle_a_2=1966092597
 constant integer unitid_gargoyle_a_3=1966092598
 constant integer unitid_gargoyle_b_1=1966092612
 constant integer unitid_gargoyle_b_2=1966092613
 constant integer unitid_gargoyle_b_3=1966092614
+constant integer unitid_gargoyle_c_1=1966092626
+constant integer unitid_gargoyle_c_2=1966092627
+constant integer unitid_gargoyle_c_3=1966092628
 constant integer unitid_gargoyle_a_1_stone=1966092599
 constant integer unitid_gargoyle_a_2_stone=1966092601
 constant integer unitid_gargoyle_a_3_stone=1966092600
 constant integer unitid_gargoyle_b_1_stone=1966092609
 constant integer unitid_gargoyle_b_2_stone=1966092610
 constant integer unitid_gargoyle_b_3_stone=1966092611
+constant integer unitid_gargoyle_c_1_stone=1966092629
+constant integer unitid_gargoyle_c_2_stone=1966092630
+constant integer unitid_gargoyle_c_3_stone=1966092631
 constant integer abilityid_stoneform_a_1=1093750338
 constant integer abilityid_stoneform_a_2=1093750339
 constant integer abilityid_stoneform_a_3=1093750340
 constant integer abilityid_stoneform_b_1=1093750348
 constant integer abilityid_stoneform_b_2=1093750349
 constant integer abilityid_stoneform_b_3=1093750350
+constant integer abilityid_stoneform_c_1=1093814618
+constant integer abilityid_stoneform_c_2=1093814832
+constant integer abilityid_stoneform_c_3=1093814833
 constant integer stoneform_stun=1093750342
 constant integer gargole_damagemax=7
 constant integer gargole_damageregen=15
 constant integer abilityid_soulassumption=1093750337
 constant integer soulassumption_indicator_0=1093750577
 constant integer soulassumption_indicator_1=1093750578
 constant integer soulassumption_indicator_2=1093750362
 constant integer soulassumption_indicator_3=1093750329
 constant integer soulassumption_indicator_4=1093750361
 constant integer soulassumption_indicator_5=1093750576
 constant integer soulassumption_indicator_6=1093750579
 constant integer soulassumption_indicator_3_full=1093750583
 constant integer soulassumption_indicator_4_full=1093750584
 constant integer soulassumption_indicator_5_full=1093750585
 constant string soulassumption_fx="Abilities\\Spells\\Undead\\AntiMagicShell\\AntiMagicShell.mdl"
@@ -4591,30 +4659,267 @@
 constant integer spellsteal_real2=1093809991
 constant integer spellsteal_real3=1093810006
 constant integer abilityid_nullfield=1093810006
 constant integer nullfield_aura=1093810005
 constant integer buffid_nullfield=1110458960
 constant integer nullfield_bonus1=1093810003
 constant integer nullfield_bonus2=1093810001
 constant integer nullfield_bonus3=1093810002
 constant integer nullfield_bonus4=1093810004
 integer nullfield_level
 constant integer abilityid_exorcism_1=1093678899
 constant integer abilityid_exorcism_2=1093677898
 constant integer abilityid_exorcism_3=1093678154
 constant integer abilityid_exorcism_4=1093678157
 constant integer abilityid_exorcism_5=1093678158
+constant integer unitid_raijin=1211117651
+constant integer abilityid_retaliation=1093810518
+constant integer abilityid_retaliation_secondary=1093810519
+constant integer buffid_retaliation=1110459734
+constant integer powerup_retaliation=1227902776
+constant string retaliation_fx_shell="war3mapImported\\RetaliationShield.mdx"
+constant string retaliation_fx_second="war3mapImported\\RetaliationTarget.mdx"
+constant integer abilityid_firebarrier=1093814355
+constant integer firebarrier_barrier=1093814325
+constant integer firebarrier_speed=1093814327
+constant integer firebarrier_speed_buffid=1110459983
+constant integer firebarrier_barrier_buffid1=1110459984
+constant integer firebarrier_barrier_buffid2=1110459985
+constant integer firebarrier_damage_1=1093814616
+constant integer firebarrier_damage_2=1093814615
+constant integer firebarrier_damage_3=1093814614
+constant integer firebarrier_damage_4=1093814617

+constant integer abilityid_slightoffist=1093814320
+constant integer slightoffist_bonus_as=1093814329
+constant integer slightoffist_bonus_1=1093814856
+constant integer slightoffist_bonus_2=1093814855
+constant integer slightoffist_bonus_3=1093814858
+constant integer slightoffist_bonus_4=1093814857
+constant integer slightoffist_bonus_buffid_dmg=1110459980
+constant integer slightoffist_bonus_buffid_as=1110459981
+constant integer slightoffist_transform_in=1093814321
+constant integer slightoffist_transform_out=1093814322
+constant integer slightoffist_creepreduction=1093814854
+constant integer abilityid_fireshackles=1093814323
+constant integer fireshackles_ensnare=1093814324
+constant integer unitid_fireshackles=1747993904
+constant integer fireremnantpath_duration=10
+constant integer abilityid_fireremnant=1093814859
+constant integer abilityid_fireremnant_secondary=1093814860

+constant string fireremnant_fx_hero="war3mapImported\\Phoenix_Missile_smaller.mdx"
+constant string fireremnant_fx_ground="war3mapImported\\FlameDash_Ground.mdx"
+constant string fireremnant_fx_impact="Abilities\\Weapons\\LordofFlameMissile\\LordofFlameMissile.mdl"
+constant integer fireremnant_dummyunit=1747993912
+integer array abilityid_fireremnant_place
+constant integer abilityid_arcaneseal=1093812536
+constant integer arcaneseal_silence=1093812535
+constant integer abilityid_thrust=1093812537
+constant integer abilityid_conjureforce=1093812534

+constant integer conjureforce_entangle=1093812549
+constant string conjureforce_fx="war3mapImported\\ConjureForceAoE.mdx"
+constant string conjureforce_fx_target="war3mapImported\\ConjureForceTarget.mdx"
+constant integer abilityid_divinewrath=1093812545
+constant integer abilityid_divinewrath_secondary=1093812546
+constant integer unitid_divinewrath=1747993673

+constant string divinewrath_fx_source="war3mapImported\\WispShellBlue_2.mdx"
+constant string divinewrath_fx="war3mapImported\\DivineWrathTarget.mdx"
+constant integer abilityid_wraith=1093812803
+constant integer abilityid_wraith_end=1093812804
+constant integer wraith_ms_buffid=1110459954
+constant integer wraith_ghost=1097167980
+constant integer wrath_magicbonus_1=1093812787
+constant integer wrath_magicbonus_2=1093812784
+constant integer wrath_magicbonus_3=1093812785
+constant integer wrath_magicbonus_4=1093812786
+constant integer abilityid_blackjack=1093812788
+constant integer abilityid_miasma=1093812789
+constant integer miasma_reveal_orderid=852625
+constant integer miasma_reveal=1093812791
+constant integer abilityid_brotherhood=1093814066
+constant integer brotherhood_fx=1093814607
+constant integer abilityid_gallop=1093812822
+constant integer gallop_ms_1=1093814604
+constant integer gallop_ms_2=1093814605
+constant integer gallop_ms_3=1093812825
+constant integer gallop_ms_4=1093814606
+constant integer gallop_ms_buffid=1110459956
+constant integer abilityid_momentofcourage=1093813593
+constant integer momentofcourage_bonus_1=1093814084
+constant integer momentofcourage_bonus_2=1093814085
+constant integer momentofcourage_bonus_3=1093814086
+constant integer momentofcourage_bonus_4=1093814083
+constant integer momentofcourage_bonus_buffid=1110459978
+constant integer momentofcourage_bonus_lifesteal_buffid=1110459979
+constant integer abilityid_duel=1093813065
+constant string duel_fx_source="war3mapImported\\DuelFX.mdx"
+constant string duel_fx_target="war3mapImported\\DuelFX.mdx"
+constant integer dual_doom=1093814069
+constant integer dual_doom_buffid=1110459975
+integer array duel_damagebonus_ability
+constant integer abilityid_deflection=1093814609
+constant string deflection_castfx="Abilities\\Spells\\Orc\\SpiritLink\\SpiritLinkZapTarget.mdl"
+constant string deflection_bufffx="Abilities\\Spells\\Orc\\SpiritLink\\SpiritLinkTarget.mdl"
+constant integer abilityid_presstheattack=1093814834
+constant string presstheattack_bufffx="Abilities\\Spells\\Human\\InnerFire\\InnerFireTarget.mdl"

+constant integer abilityid_cleanse=1093814850
+constant string cleanse_fx="Abilities\\Spells\\Human\\DispelMagic\\DispelMagicTarget.mdl"
+constant integer cleanse_ms=1093814851
+constant integer cleanse_ms_buffid=1110459991

+constant integer abilityid_starbeam=1093812807
+constant string starbeam_fx="war3mapImported\\DivineWrathTarget.mdx"
+unit starbeam_source
+integer starbeam_count
+integer starbeam_level
+constant integer abilityid_amp=1093812820
+constant string amp_fx="Abilities\\Spells\\Undead\\ReplenishMana\\SpiritTouchTarget.mdl"
+constant string amp_fx_initial="war3mapImported\\FarseerMissile_Purple_2.mdx"
+constant integer amp_slow1=1093814582
+constant integer amp_slow2=1093814581
+constant integer amp_slow3=1093814583
+constant integer amp_slow4=1093814584
+constant integer amp_slow_buffid=1110459987
+boolean amp_allow=false
+constant integer abilityid_arcanebolt=1093812805
+constant integer arcanebolt_unitid=1747993681
+constant integer abilityid_arcanesilence=1093814350
+constant string arcanesilence_fx="Abilities\\Spells\\Undead\\ReplenishMana\\SpiritTouchTarget.mdl"
+constant integer arcanesilence_amp_1=1093814360
+constant integer arcanesilence_amp_2=1093814576
+constant integer arcanesilence_amp_3=1093814361
+constant integer arcanesilence_amp_4=1093814362
+constant integer abilityid_concussiveshot=1093814612
+constant integer unitid_concussiveshot=1747993910
+constant integer concussiveshot_slow=1093814613
+constant integer abilityid_deathwatch=1093813335
+constant string deathwatch_fx="war3mapImported\\GaiaShield.mdx"
+constant integer abilityid_realityshift=1093813336
+constant string realityshift_fx=""
+constant integer unitid_realityshift_1=1865429330
+constant integer unitid_realityshift_2=1865429331
+constant integer unitid_realityshift_3=1865429329
+constant integer unitid_realityshift_4=1865429332
+constant integer realityshift_banish=1093813337

+constant integer abilityid_hauntingechos=1093813552
+
+constant integer unitid_hauntingechos=1747993682
+constant integer hauntingechos_silence=1093813553

+constant integer abilityid_fatalattraction=1093813338
+constant string fatalattraction_fx=""
+constant integer abilityid_whirlingdeath=1093813835
+constant real whirlingdeath_aoe=300
+constant string whirlingdeath_fx="war3mapImported\\WhirlingDeath_01.mdx"
+integer whirlingdeath_type
+constant integer abilityid_timberchain=1093813555
+constant integer unitid_timberchain=1966092625
+constant integer unitid_timberchainhead=1966092624
+group timberchain_group
+integer timberchain_count
+constant integer abilityid_reactivearmor=1093813556
+integer array reactivearmor_bonus
+constant integer abilityid_chakram=1093813557
+constant integer abilityid_chakram_return=1093813848
+constant integer chakram_disableattack=1096971630
+constant real chakram_aoe=200
+constant real chakram_dps_aoe=200
+constant integer unitid_chakram=1747993683
+constant integer chakram_slow=1093813334
+group chakram_group
+unit chakram_source
+integer chakram_level
+constant integer abilityid_breeze=1093813582
+constant integer unitid_breeze=1747993688
+constant integer breeze_slow_1=1093813580
+constant integer breeze_slow_2=1093813579
+constant integer breeze_slow_3=1093813578
+constant integer breeze_slow_4=1093813581
+constant integer breeze_slow_buffid=1110459958
+unit breeze_source
+unit breeze_projectile
+group breeze_group
+constant integer abilityid_frostskin=1093813593
+constant integer frostskin_reduction_1=1093813592
+constant integer frostskin_reduction_2=1093813589
+constant integer frostskin_reduction_3=1093813590
+constant integer frostskin_reduction_4=1093813591
+constant integer frostskin_mana_1=1093813587
+constant integer frostskin_mana_2=1093813586
+constant integer frostskin_mana_3=1093813585
+constant integer frostskin_mana_4=1093813588
+constant integer abilityid_frostgiant=1093813594
+constant integer frostgiant_sourceid=1311788343
+constant integer unitid_frostgiant=1865429333
+constant integer abilityid_sonicboom=1093813812
+constant string sonicboom_impactfx="war3mapImported\\BlackDragonMissile.mdx"
+unit sonicboom_source
+integer sonicboom_level
+constant integer abilityid_windblast=1093813573
+constant integer windblast_armor=1093813811
+constant string windblast_fx="effects\\BasicWaterFlash.mdx"
+unit windblast_source
+integer windblast_level
+constant integer abilityid_windbarrier=1093813808
+constant integer windbarrier_speed=1093813810
+constant integer windbarrier_speed_buffid=1110459959
+constant integer windbarrier_barrier_buffid1=1110456374
+constant integer windbarrier_barrier_buffid2=1110456375
+constant integer abilityid_prevailingwinds=1093813813
+constant integer prevailingwinds_cyclone=1093813833
+constant integer prevailingwinds_slow_1=1093813828
+constant integer prevailingwinds_slow_2=1093813827
+constant integer prevailingwinds_slow_3=1093813829
+constant integer prevailingwinds_slow_buffid=1110459961
+constant integer prevailingwinds_haste_1=1093813831
+constant integer prevailingwinds_haste_2=1093813830
+constant integer prevailingwinds_haste_3=1093813832
+constant integer prevailingwinds_haste_buffid=1110459960
+constant integer abilityid_holyshackles=1093813844
+constant integer holyshackles_ensnare=1093813845
+constant integer unitid_holyshackles=1747993690
+constant integer abilityid_focus=1093813841
+constant integer focus_bonus=1093813838
+constant integer focus_bonus_buffid=1110459969
+constant integer abilityid_divineintervention=1093813836
+constant integer divineintervention_invul=1098282348
+constant string divineintervention_fx="Abilities\\Spells\\Human\\DivineShield\\DivineShieldTarget.mdl"
+constant integer divineintervention_doom=1093813847
+constant integer divineintervention_disableattack=1096971630
+constant integer abilityid_chill=1093813583
+constant integer chill_silence=1093813584
 endglobals
 function InitGlobals takes nothing returns nothing
 local integer i=0
 set i=0
 loop
 exitwhen(i>16)
 set udg_RevivalTimer[i]=0
 set i=i+1
 endloop
 set i=0
 loop
 exitwhen(i>100)
 set udg_HeroFlags[i]=false
 set i=i+1
 endloop
@@ -35334,30 +36704,36 @@
 endif
 endif
 endfunction
 function Command_NoSwitch takes nothing returns nothing
 if udg_Elapsed15==false and switch_enabled and GetTriggerPlayer()==udg_Host then
 set switch_enabled=false
 call DisplayTimedTextToPlayer(GetLocalPlayer(),0,0,10,GetObjectName(MSG_SWITCH_15))
 endif
 endfunction
 function Command_Sleep takes nothing returns nothing
 call SetUnitAnimation(udg_Hero[GetPlayerId(GetTriggerPlayer())],"sleep")
 endfunction
 function Command_Calm takes nothing returns nothing
 call SetUnitAnimationByIndex(udg_Hero[GetPlayerId(GetTriggerPlayer())],23)
 endfunction
+function Command_NoHat takes nothing returns nothing
+if udg_Hero[GetPlayerId(GetTriggerPlayer())]!=null then
+set nohat_setting[GetPlayerId(GetTriggerPlayer())]=1
+call UnitRemoveAbility(udg_Hero[GetPlayerId(GetTriggerPlayer())],1093815114)
+endif
+endfunction
 function Command_Tips takes nothing returns nothing
 if tips_enabled[GetPlayerId(GetTriggerPlayer())]==false then
 set tips_enabled[GetPlayerId(GetTriggerPlayer())]=true
 if GetObjectName(1848657992)=="Language: English"then
 call DisplayTimedTextToPlayer(GetTriggerPlayer(),0,0,10,"You have activated the -tips system.")
 call DisplayTimedTextToPlayer(GetTriggerPlayer(),0,0,10,"Please remember that these are just introductory recommendations and are not the only way to play heroes.")
 endif
 if udg_Hero[GetPlayerId(GetTriggerPlayer())]!=null then
 call InitiateTips(GetTriggerPlayer(),GetUnitPointValue(udg_Hero[GetPlayerId(GetTriggerPlayer())]))
 endif
 endif
 endfunction
 function InitTrig_Tips takes nothing returns nothing
 set tips_enabled[0]=false
 set tips_enabled[1]=false
@@ -42922,50 +44298,51 @@
 set Target=null
 return false
 endfunction
 function Rum_Phase1 takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local unit Target=(LoadUnitHandle(hash_main,(Cache),(17)))
 local player p=(LoadPlayerHandle(hash_main,(Cache),(54)))
 local trigger Phase2=(LoadTriggerHandle(hash_main,(Cache),(135)))
 local integer Index=GetTriggerEvalCount(t)
 if GetTriggerEventId()!=EVENT_UNIT_DAMAGED then
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
 call TriggerRegisterTimerEvent(Phase2,1,true)
 call TriggerAddCondition(Phase2,Condition(function Rum_Phase2))
+call UnitRemoveAbility(Target,buffid_rum_ms)
 else
 set Cache=GetHandleId(Phase2)
 call SetUnitState(Target,UNIT_STATE_LIFE,GetUnitState(Target,UNIT_STATE_LIFE)+0.5*GetEventDamage())
 call SavePlayerHandle(hash_main,(Cache),(1300+Index),(GetOwningPlayer(GetEventDamageSource())))
 call SaveReal(hash_main,(Cache),(1200+Index),((0.5*GetEventDamage())*1.0))
 call SaveInteger(hash_main,(Cache),(136),(Index))
 endif
 set t=null
 set Target=null
 set p=null
 set Phase2=null
 return false
 endfunction
 function GhostShip_Hit takes nothing returns nothing
 local unit Source=ghostship_source
 local unit Target=GetEnumUnit()
 local integer Level=ghostship_level
 local trigger t=CreateTrigger()
 local integer Cache=GetHandleId(t)
-local integer Duration=4+Level
+local integer Duration=8
 local trigger Phase2=CreateTrigger()
 call AddTimedAbility(Target,rum_bonus,1,Duration)
 call UnitMakeAbilityPermanent(Target,true,rum_bonus)
 call SetPlayerAbilityAvailable(GetOwningPlayer(Target),rum_bonus,false)
 call TriggerRegisterTimerEvent(t,Duration+0.01,false)
 call TriggerRegisterUnitEvent(t,Target,EVENT_UNIT_DAMAGED)
 call TriggerAddCondition(t,Condition(function Rum_Phase1))
 call SavePlayerHandle(hash_main,(Cache),(54),(GetOwningPlayer(Source)))
 call SaveUnitHandle(hash_main,(Cache),(17),(Target))
 call SaveTriggerHandle(hash_main,(Cache),(135),(Phase2))
 call SaveUnitHandle(hash_main,(GetHandleId(Phase2)),(17),(Target))
 call SaveInteger(hash_main,(GetHandleId(Phase2)),(188),(Duration))
 set Source=null
 set Target=null
 set t=null
@@ -43055,46 +44432,50 @@
 local real SourceY=GetUnitY(Hero)
 local location l=GetSpellTargetLoc()
 local real x0=GetLocationX(l)
 local real y0=GetLocationY(l)
 local real Angle
 local real x1
 local real y1
 local real x2
 local real y2
 local trigger t
 local integer Cache
 local integer Level
 local unit Projectile
 local unit Projectile2
 local string s=""
+local integer shipid=unitid_ghostship
+if GetRandomInt(0,5)==0 then
+set shipid=unitid_ghostship_alt
+endif
 if IsPlayerAlly(GetOwningPlayer(Hero),GetLocalPlayer())or IsObserver(GetLocalPlayer())then
 set s="war3mapImported\\Whirlpool.mdx"
 endif
 if x0==GetUnitX(Hero)and y0==GetUnitY(Hero)then
 set x0=x0+50*Cos(GetUnitFacing(Hero)*bj_DEGTORAD)
 set y0=y0+50*Sin(GetUnitFacing(Hero)*bj_DEGTORAD)
 endif
 set Angle=Atan2(y0-SourceY,x0-SourceX)
 set x1=SourceX-1000*Cos(Angle)
 set y1=SourceY-1000*Sin(Angle)
 set x2=SourceX+1000*Cos(Angle)
 set y2=SourceY+1000*Sin(Angle)
 set t=CreateTrigger()
 set Cache=GetHandleId(t)
 set Level=GetUnitAbilityLevel(Hero,abilityid_ghostship)
-set Projectile=CreateUnit(GetOwningPlayer(Hero),unitid_ghostship,x1,y1,Atan2(y0-SourceY,x0-SourceX)*bj_RADTODEG)
+set Projectile=CreateUnit(GetOwningPlayer(Hero),shipid,x1,y1,Atan2(y0-SourceY,x0-SourceX)*bj_RADTODEG)
 set Projectile2=CreateUnit(GetOwningPlayer(Hero),unitid_ghostship2,x1,y1,Atan2(y0-SourceY,x0-SourceX)*bj_RADTODEG)
 call SetUnitVertexColor(Projectile,255,255,255,0)
 call SetUnitPathing(Projectile,false)
 call RemoveLocation(l)
 call SaveInteger(hash_main,(Cache),(5),(Level))
 call SaveReal(hash_main,(Cache),(13),((Angle)*1.0))
 call SaveUnitHandle(hash_main,(Cache),(45),(Projectile))
 call SaveUnitHandle(hash_main,(Cache),(186),(Projectile2))
 call SaveReal(hash_main,(Cache),(66),((x1)*1.0))
 call SaveReal(hash_main,(Cache),(67),((y1)*1.0))
 call SaveGroupHandle(hash_main,(Cache),(187),(GetAvailableGroup()))
 call DestroyEffect(AddSpecialEffect(s,SafeX(x2),SafeY(y2)))
 call TriggerRegisterTimerEvent(t,0.02,true)
 call TriggerAddCondition(t,Condition(function GhostShip_Move))
 set Hero=null
@@ -43343,31 +44724,31 @@
 local unit u=(LoadUnitHandle(hash_main,(Cache),(26)))
 call SaveInteger(hash_main,(GetHandleId(u)),(27),((LoadInteger(hash_main,(GetHandleId(u)),(27)))-1))
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
 set u=null
 set t=null
 return false
 endfunction
 function GoblinsGreed_Action takes nothing returns nothing
 local unit UnitVar=GetKillingUnit()
 local unit DeadUnit=GetTriggerUnit()
 local integer Cache=GetHandleId(UnitVar)
 local integer PreviousKills=(LoadInteger(hash_main,(Cache),(27)))
 local integer Level=GetUnitAbilityLevel(UnitVar,1093685043)
 local integer ExtraBonusGold=PreviousKills*2
-local integer TotalBonusGold=IMin(ExtraBonusGold+Level*2,18)
+local integer TotalBonusGold=IMin(ExtraBonusGold+Level*2,26)
 local texttag tt=CreateTextTag()
 local player p=GetOwningPlayer(UnitVar)
 local trigger t=CreateTrigger()
 set GG_Bonus[GetPlayerId(p)]=GG_Bonus[GetPlayerId(p)]+TotalBonusGold
 call TriggerRegisterTimerEvent(t,20,false)
 call TriggerAddCondition(t,Condition(function GoblinsGreed_Decrement))
 call SaveUnitHandle(hash_main,(GetHandleId(t)),(26),(UnitVar))
 call SaveInteger(hash_main,(Cache),(27),(PreviousKills+1))
 call SetPlayerState(p,PLAYER_STATE_RESOURCE_GOLD,TotalBonusGold+GetPlayerState(p,PLAYER_STATE_RESOURCE_GOLD))
 call SetTextTagText(tt,"+"+I2S(TotalBonusGold),0.025)
 call SetTextTagPosUnit(tt,DeadUnit,0)
 call SetTextTagColor(tt,255,220,0,255)
 call SetTextTagVelocity(tt,0,0.03)
 if GetLocalPlayer()==p then
 call SetTextTagVisibility(tt,true)
@@ -47389,40 +48869,36 @@
 set Hero=null
 endfunction
 function WanningRift_Main takes nothing returns boolean
 if GetSpellAbilityId()==abilityid_wanningrift then
 call WanningRift()
 endif
 return false
 endfunction
 function Register_WanningRift takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
 call TriggerAddCondition(t,Condition(function WanningRift_Main))
 set t=null
 endfunction
 function PrismaticPrison_MainStun takes unit Source,unit Target,integer Level,integer Type returns nothing
-local unit Caster=CreateUnit(GetOwningPlayer(Source),1697656901,GetUnitX(Target),GetUnitY(Target),0)
-call UnitAddAbility2(Caster,abilityid_prismaticprisonstunmain)
-call SetUnitAbilityLevel(Caster,abilityid_prismaticprisonstunmain,Level)
-call IssueTargetOrder(Caster,"thunderbolt",Target)
+call Stun(Source,Target,0.75+0.75*Level)
 if Type==1 then
 call DamageTarget(Source,Target,DAMAGE_MAGICAL,Level*50+50)
 else
 call DamageTarget(Source,Target,DAMAGE_MAGICAL,Level*50+50+100)
 endif
-set Caster=null
 endfunction
 function PrismaticPrison_MiniStun takes unit Source,unit Target returns nothing
 local unit Caster=CreateUnit(GetOwningPlayer(Source),1697656901,GetUnitX(Target),GetUnitY(Target),0)
 call UnitAddAbility2(Caster,abilityid_prismaticprisonstun)
 call IssueTargetOrder(Caster,"thunderbolt",Target)
 set Caster=null
 endfunction
 function PrismaticPrison_Iterate takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local integer Level=(LoadInteger(hash_main,(Cache),(5)))
 local integer Type=(LoadInteger(hash_main,(Cache),(152)))
 local unit Source=(LoadUnitHandle(hash_main,(Cache),(2)))
 local integer TargetIndex=(LoadInteger(hash_main,(Cache),(30)))
 local unit Target=TI_GetUnit(TargetIndex)
@@ -50058,31 +51534,31 @@
 local real OriginalX
 local real OriginalY
 call SetUnitX(Projectile,NewX)
 call SetUnitY(Projectile,NewY)
 set udg_TempUnit=Hero
 call GroupEnumUnitsInRange(GroupVar,NewX,NewY,140,CondVar)
 set Target=FirstOfGroup(GroupVar)
 call KillGroup(GroupVar)
 if Target!=null then
 if IsUnitType(Target,UNIT_TYPE_HERO)==true then
 set Arrow_Hits[GetPlayerId(GetOwningPlayer(Hero))]=Arrow_Hits[GetPlayerId(GetOwningPlayer(Hero))]+1
 call Traffic_RegisterData("AA_Hits"+I2S(GetPlayerId(GetOwningPlayer(Hero))),Arrow_Hits[GetPlayerId(GetOwningPlayer(Hero))])
 endif
 set OriginalX=(LoadReal(hash_main,(Cache),(191)))
 set OriginalY=(LoadReal(hash_main,(Cache),(192)))
-set Caster=CreateUnit(GetOwningPlayer(Hero),1697656901,SourceX,SourceY,0)
+set Caster=CreateUnit(GetOwningPlayer(Target),1697656901,SourceX,SourceY,0)
 call UnitAddAbility2(Caster,(1093684279))
 call SetUnitAbilityLevel(Caster,(1093684279),ElunesArrow_GetStunLevel(OriginalX,OriginalY,SourceX,SourceY))
 if IssueTargetOrder(Caster,"thunderbolt",Target)then
 endif
 call UnitDamageTarget(Hero,Target,90*GetUnitAbilityLevel(Hero,(1093684280)),true,false,ATTACK_TYPE_NORMAL,DAMAGE_TYPE_MAGIC,WEAPON_TYPE_WHOKNOWS)
 call KillUnit(Projectile)
 call KillTimer(t)
 elseif(NewX-TargetX)*(NewX-TargetX)+(NewY-TargetY)*(NewY-TargetY)<1600 then
 call KillUnit(Projectile)
 call KillTimer(t)
 endif
 endfunction
 function ElunesArrow_Action takes nothing returns nothing
 local unit Hero=GetTriggerUnit()
 local real SourceX=SafeX(GetUnitX(Hero))
@@ -51068,30 +52546,32 @@
 endfunction
 function MC_BL_WhichUnits takes nothing returns boolean
 local unit FilterUnit=GetFilterUnit()
 local unit UnitVar=GetTriggerUnit()
 if IsUnitAlly(FilterUnit,GetOwningPlayer(UnitVar))and GetUnitState(FilterUnit,UNIT_STATE_LIFE)>0 and GetUnitAbilityLevel(FilterUnit,1113746543)==0 then
 if(GetOwningPlayer(FilterUnit)==udg_SentPlayers[0]or GetOwningPlayer(FilterUnit)==udg_ScourgePlayers[0])then
 if(GetUnitTypeId(FilterUnit)==1969711215 or GetUnitTypeId(FilterUnit)==1966092337 or GetUnitTypeId(FilterUnit)==1970169187 or GetUnitTypeId(FilterUnit)==1966092338)then
 return true
 endif
 if(GetUnitTypeId(FilterUnit)==1702061422 or GetUnitTypeId(FilterUnit)==1697656918 or GetUnitTypeId(FilterUnit)==1701081721 or GetUnitTypeId(FilterUnit)==1697656919)then
 return true
 endif
 return false
 elseif IsUnitType(FilterUnit,UNIT_TYPE_HERO)==true then
 return true
+elseif GetOwningPlayer(FilterUnit)==GetOwningPlayer(UnitVar)then
+return true
 endif
 endif
 return false
 endfunction
 function Bloodlust takes integer r returns nothing
 local unit UnitVar=GetTriggerUnit()
 local integer Level=r
 local group GroupVar=GetAvailableGroup()
 local unit Caster
 local unit TempUnit
 local integer SpellId=GetSpellAbilityId()
 local integer loopcurrent=1
 if r==2 then
 call TextTagOnUnitEx(GetObjectName(1848658743),5,GetTriggerUnit(),0.03,255,0,0,255)
 elseif r==3 then
@@ -51331,31 +52811,31 @@
 set t=null
 endfunction
 function Juxtapose_IsImage takes nothing returns boolean
 if GetUnitAbilityLevel(GetFilterUnit(),buffid_juxtapose)>0 then
 set tt_int1=tt_int1+1
 endif
 return false
 endfunction
 function Juxtapose takes nothing returns nothing
 local unit Lancer=GetAttacker()
 local group g=GetAvailableGroup()
 local integer Rand=GetRandomInt(1,100)
 local integer JuxLevel=GetUnitAbilityLevel(Lancer,abilityid_juxtapose)
 local integer EdgeLevel=GetUnitAbilityLevel(Lancer,abilityid_phantomedge)
 local unit Caster
-local integer MaxIllusions=2+JuxLevel
+local integer MaxIllusions=2*JuxLevel
 local integer JuxChance=12
 set tt_int1=0
 call GroupEnumUnitsOfPlayer(g,GetOwningPlayer(Lancer),Condition(function Juxtapose_IsImage))
 call KillGroup(g)
 if tt_int1<MaxIllusions then
 if IsUnitIllusion(Lancer)==false and Rand<=(JuxChance+2*EdgeLevel)then
 set Caster=CreateUnit(GetOwningPlayer(Lancer),1697656901,GetUnitX(Lancer),GetUnitY(Lancer),0)
 call UnitAddAbility2(Caster,abilityid_juxtaposeimage)
 call IssueTargetOrderById(Caster,852274,Lancer)
 endif
 if IsUnitIllusion(Lancer)==true and Rand<=(1+2*EdgeLevel)and EdgeLevel>0 then
 set Caster=CreateUnit(GetOwningPlayer(Lancer),1697656901,GetUnitX(Lancer),GetUnitY(Lancer),0)
 call UnitAddAbility2(Caster,abilityid_juxtaposeimage)
 call IssueTargetOrderById(Caster,852274,Lancer)
 endif@@ -52225,31 +53705,31 @@
 call VolumeGroupSetVolume(SOUND_VOLUMEGROUP_MUSIC,.0)
 call RemoveLocation(PointVar)
 set t=null
 set PointVar=null
 endfunction
 function GS takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
 call TriggerAddCondition(t,Condition(function Trig_Global_Silence_Conditions))
 call TriggerAddAction(t,function Trig_Global_Silence_Actions)
 endfunction
 function GlaivesOfWisdom_Attack_Damage takes nothing returns boolean
 local real d
 if GetUnitAbilityLevel(GetTriggerUnit(),buffid_glaivesofwisdom)>0 and GetUnitAbilityLevel(GetEventDamageSource(),abilityid_glaivesofwisdom)>0 and(LoadUnitHandle(hash_main,(GetHandleId(GetTriggeringTrigger())),(2)))==GetEventDamageSource()then
 call UnitRemoveAbility(GetTriggerUnit(),buffid_glaivesofwisdom)
-set d=(0.2+0.1*I2R(GetUnitAbilityLevel(GetEventDamageSource(),abilityid_glaivesofwisdom)))*GetHeroInt(GetEventDamageSource(),true)
+set d=(0.15+0.15*I2R(GetUnitAbilityLevel(GetEventDamageSource(),abilityid_glaivesofwisdom)))*GetHeroInt(GetEventDamageSource(),true)
 call TextTagOnUnitEx("+"+I2S(R2I(d)),1,GetTriggerUnit(),0.023,3,216,216,216)
 call DisableTrigger(GetTriggeringTrigger())
 call DamageTarget(GetEventDamageSource(),GetTriggerUnit(),DAMAGE_PURE,d)
 call FlushChildHashtable(hash_main,(GetHandleId(GetTriggeringTrigger())))
 call CleanTrigger(GetTriggeringTrigger())
 endif
 return false
 endfunction
 function GlaivesOfWisdom_Attack_Child takes nothing returns nothing
 local trigger t
 local unit Target
 local unit Source
 if GetTriggerEventId()==EVENT_UNIT_SPELL_EFFECT then
 set Target=GetSpellTargetUnit()
 set Source=GetTriggerUnit()
@@ -52396,31 +53876,31 @@
 endfunction
 function Shrapnel_Iterate takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local unit Source=(LoadUnitHandle(hash_main,(Cache),(2)))
 local unit Caster=(LoadUnitHandle(hash_main,(Cache),(19)))
 local integer Level=(LoadInteger(hash_main,(Cache),(5)))
 local real x=(LoadReal(hash_main,(Cache),(6)))
 local real y=(LoadReal(hash_main,(Cache),(7)))
 local group g=GetAvailableGroup()
 local real d
 local real a
 local real x2
 local real y2
 set tt_unit1=Source
-call GroupEnumUnitsInRange(g,x,y,350,Condition(function GenericCondition_EnemyUnitsAndStructureEx))
+call GroupEnumUnitsInRange(g,x,y,375,Condition(function GenericCondition_EnemyUnitsAndStructureEx))
 set shrapnel_source=Source
 set shrapnel_damage=Level*12
 set shrapnel_caster=Caster
 call ForGroup(g,function Shrapnel_Hit)
 call KillGroup(g)
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(309))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(175))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(176))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(177))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(178))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(179))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(180))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(330))))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(331))))
 if GetTriggerEvalCount(t)==8 then
@@ -52710,53 +54190,53 @@
 set t=null
 endfunction
 function Backstab takes nothing returns nothing
 local unit Source=GetAttacker()
 local unit Target=GetTriggerUnit()
 local integer Level=GetUnitAbilityLevel(Source,abilityid_backstab)
 local boolean Free=(LoadBoolean(hash_main,(GetHandleId(Source)),(332)))
 local real d=RAbsBJ(GetUnitFacing(Target)-GetUnitFacing(Source))
 if d>180 then
 set d=360-d
 endif
 if d<=105 or Free then
 call SaveBoolean(hash_main,(GetHandleId(Source)),(332),(false))
 if GetUnitAbilityLevel(Target,1093678162)==0 then
 call DestroyEffect(AddSpecialEffectTarget("Abilities\\Spells\\Other\\Stampede\\StampedeMissileDeath.mdl",Target,"chest"))
-call DamageTarget(Source,Target,DAMAGE_PHYSICAL,GetHeroAgi(Source,true)*0.25*Level)
+call DamageTarget(Source,Target,DAMAGE_PHYSICAL,GetHeroAgi(Source,true)*(0.25*Level+0.25))
 endif
 endif
 set Source=null
 set Target=null
 endfunction
 function Backstab_Main takes nothing returns boolean
 if GetUnitAbilityLevel(GetAttacker(),abilityid_backstab)>0 and IsUnitType(GetTriggerUnit(),UNIT_TYPE_STRUCTURE)==false and IsUnitAlly(GetTriggerUnit(),GetOwningPlayer(GetAttacker()))==false then
 call Backstab()
 endif
 return false
 endfunction
 function Register_Backstab takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_ATTACKED)
 call TriggerAddCondition(t,Condition(function Backstab_Main))
 set t=null
 endfunction
 function Warcry_Add takes nothing returns nothing
 if IsRealSummoned(GetEnumUnit())then
 return
 endif
-call AddTimedAbilityAndRemove(GetEnumUnit(),warcry_abilityid,1,6,1110458966)
+call AddTimedAbilityAndRemove(GetEnumUnit(),warcry_abilityid,1,8,1110458966)
 call UnitMakeAbilityPermanent(GetEnumUnit(),true,warcry_abilityid)
 call SetPlayerAbilityAvailable(GetOwningPlayer(GetEnumUnit()),warcry_abilityid,false)
 endfunction
 function Warcry takes nothing returns nothing
 local unit Source=GetTriggerUnit()
 local integer Level=GetUnitAbilityLevel(Source,abilityid_warcry)
 local group g=GetAvailableGroup()
 if Level==1 then
 set warcry_abilityid=warcry_abilityid_1
 elseif Level==2 then
 set warcry_abilityid=warcry_abilityid_2
 elseif Level==3 then
 set warcry_abilityid=warcry_abilityid_3
 elseif Level==4 then
 set warcry_abilityid=warcry_abilityid_4
@@ -52778,37 +54258,39 @@
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
 call TriggerAddCondition(t,Condition(function Warcry_Main))
 set t=null
 endfunction
 function GodStrength_End takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local unit Hero=(LoadUnitHandle(hash_main,(Cache),(14)))
 local integer Count=(LoadInteger(hash_main,(Cache),(34)))
 if GetTriggerEventId()!=EVENT_PLAYER_UNIT_SPELL_EFFECT then
 set Count=Count+1
 call SaveInteger(hash_main,(Cache),(34),(Count))
 endif
 if(GetTriggerEventId()!=EVENT_PLAYER_UNIT_SPELL_EFFECT and Count>1)or(GetTriggerEventId()==EVENT_PLAYER_UNIT_SPELL_EFFECT and IsPurge(GetSpellAbilityId())and GetSpellTargetUnit()==Hero)then
+if(GetTriggerEventId()!=EVENT_PLAYER_UNIT_SPELL_EFFECT and Count>1)then
 if IsDead(Hero)==false then
 call UnitRemoveAbility(Hero,abilityid_godstrengthbonus)
 call UnitRemoveAbility(Hero,1110454329)
 call UnitRemoveAbility(Hero,1110459472)
 endif
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
+endif
 elseif GetTriggerEventId()!=EVENT_PLAYER_UNIT_SPELL_EFFECT then
 call DestroyEffect(AddSpecialEffectTarget("war3mapImported\\BattleCryCaster.mdx",Hero,"overhead"))
 call UnitAddAbility2(Hero,abilityid_godstrengthbonus)
 call SetUnitAbilityLevel(Hero,abilityid_godstrengthbonus,GetUnitAbilityLevel(Hero,abilityid_godstrength))
 call UnitMakeAbilityPermanent(Hero,true,abilityid_godstrengthbonus)
 call SetPlayerAbilityAvailable(GetOwningPlayer(Hero),abilityid_godstrengthbonus,false)
 endif
 set t=null
 set Hero=null
 return false
 endfunction
 function GodStrength takes nothing returns nothing
 local trigger t=CreateTrigger()
 local integer Cache=GetHandleId(t)
 local unit Hero=GetTriggerUnit()
@@ -54425,31 +55907,31 @@
 return true
 endif
 if((GetUnitTypeId(GetFilterUnit())==1848651856))then
 return true
 endif
 if((GetUnitTypeId(GetFilterUnit())==1848651857))then
 return true
 endif
 if((GetUnitTypeId(GetFilterUnit())==1848651854))then
 return true
 endif
 endif
 return false
 endfunction
 function LandMine_TripHeroes takes nothing returns boolean
-return(GetUnitTypeId(GetFilterUnit())==1211117637 or GetUnitTypeId(GetFilterUnit())==1211117639 or GetUnitTypeId(GetFilterUnit())==1211117638 or GetUnitTypeId(GetFilterUnit())==1160786000 or GetUnitTypeId(GetFilterUnit())==1311780930 or GetUnitTypeId(GetFilterUnit())==1430468144 or GetUnitTypeId(GetFilterUnit())==1162032951 or GetUnitTypeId(GetFilterUnit())==1429221456 or GetUnitTypeId(GetFilterUnit())==1160786510 or GetUnitTypeId(GetFilterUnit())==1160786511 or GetUnitTypeId(GetFilterUnit())==1328558390 or GetUnitTypeId(GetFilterUnit())==1328558391 or GetUnitTypeId(GetFilterUnit())==1160786502)and IsUnitEnemy(GetFilterUnit(),GetOwningPlayer(tt_unit1))
+return(GetUnitTypeId(GetFilterUnit())==1211117637 or GetUnitTypeId(GetFilterUnit())==1211117639 or GetUnitTypeId(GetFilterUnit())==1211117638 or GetUnitTypeId(GetFilterUnit())==1160786000 or GetUnitTypeId(GetFilterUnit())==1311780930 or GetUnitTypeId(GetFilterUnit())==1430468144 or GetUnitTypeId(GetFilterUnit())==1162032951 or GetUnitTypeId(GetFilterUnit())==1429221456 or GetUnitTypeId(GetFilterUnit())==1160786510 or GetUnitTypeId(GetFilterUnit())==1160786511 or GetUnitTypeId(GetFilterUnit())==1328558390 or GetUnitTypeId(GetFilterUnit())==1328558391 or GetUnitTypeId(GetFilterUnit())==1160786502 or GetUnitTypeId(GetFilterUnit())==1211122767)and IsUnitEnemy(GetFilterUnit(),GetOwningPlayer(tt_unit1))
 endfunction
 function LandMine_Scan takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local unit Mine=(LoadUnitHandle(hash_main,(Cache),(347)))
 local group g
 if GetTriggerEventId()==EVENT_UNIT_DEATH then
 call DestroyEffect(AddSpecialEffect("war3mapImported\\NewGroundEX.mdx",GetUnitX(Mine),GetUnitY(Mine)))
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
 else
 set g=GetAvailableGroup()
 set tt_unit1=Mine
 call GroupEnumUnitsInRange(g,GetUnitX(Mine),GetUnitY(Mine),200,Condition(function LandMine_TripHeroes))
 if FirstOfGroup(g)!=null then
@@ -54608,31 +56090,31 @@
 set l=null
 endfunction
 function MarchOfTheMachines_Main takes nothing returns boolean
 if GetSpellAbilityId()==abilityid_marchofthemachines then
 call MarchOfTheMachines()
 endif
 return false
 endfunction
 function Register_MarchOfTheMachines takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
 call TriggerAddCondition(t,Condition(function MarchOfTheMachines_Main))
 set t=null
 endfunction
 function Trig_Laser_Conditions takes nothing returns boolean
-return GetSpellAbilityId()==1093678137
+return GetSpellAbilityId()==1093678137 and IsBlocked(GetSpellTargetUnit())==false
 endfunction
 function Trig_Laser_Actions takes nothing returns nothing
 call UnitDamageTarget(GetTriggerUnit(),GetSpellTargetUnit(),GetUnitAbilityLevel(GetTriggerUnit(),1093678137)*80,true,false,ATTACK_TYPE_HERO,DAMAGE_TYPE_MAGIC,WEAPON_TYPE_WHOKNOWS)
 endfunction
 function Laser takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
 call TriggerAddCondition(t,Condition(function Trig_Laser_Conditions))
 call TriggerAddAction(t,function Trig_Laser_Actions)
 endfunction
 function Toss_DealDamage takes nothing returns nothing
 if IsUnitType(GetEnumUnit(),UNIT_TYPE_STRUCTURE)==true then
 call DamageTarget(toss_source,GetEnumUnit(),DAMAGE_MAGICAL,toss_damage/3)
 else
 call DamageTarget(toss_source,GetEnumUnit(),DAMAGE_MAGICAL,toss_damage)
@@ -55181,31 +56677,31 @@
 call SetUnitAbilityLevel(GetTriggerUnit(),1093679429,GetUnitAbilityLevel(GetTriggerUnit(),1093681733))
 endfunction
 function BerserkerL takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_HERO_SKILL)
 call TriggerAddCondition(t,Condition(function Trig_Berserker_Rage_learned_Conditions))
 call TriggerAddAction(t,function Trig_Berserker_Rage_learned_Actions)
 endfunction
 constant function BeastRage_AbilityCode takes nothing returns integer
 return 1093685040
 endfunction
 constant function BeastRage_EffectCode takes nothing returns integer
 return 1093684826
 endfunction
 function BeastRage_Conditions takes nothing returns boolean
-return GetUnitAbilityLevel(GetAttacker(),(1093685040))>0 and((LoadInteger(hash_main,(GetHandleId((GetAttacker()))),((4271))))==HERO_STATE_ON)==false and IsUnitType(GetTriggerUnit(),UNIT_TYPE_STRUCTURE)==false
+return GetUnitAbilityLevel(GetAttacker(),(1093685040))>0 and((LoadInteger(hash_main,(GetHandleId((GetAttacker()))),((4271))))==HERO_STATE_ON)==false
 endfunction
 function BeastRage_Actions takes nothing returns nothing
 local unit Attacker=GetAttacker()
 local unit Target=GetTriggerUnit()
 local integer Cache=GetHandleId(Attacker)
 local integer PreviousTargetIndex=(LoadInteger(hash_main,(Cache),(237)))
 local unit PreviousTarget=TI_GetUnit(PreviousTargetIndex)
 local integer Counter=IMax((LoadInteger(hash_main,(Cache),(236))),0)
 local integer Level=GetUnitAbilityLevel(Attacker,(1093685040))
 if PreviousTarget!=Target then
 set Counter=1
 else
 set Counter=IMin(Counter+1,2+Level)
 endif
 call AddAbilityLevel(Attacker,(1093684826),Counter)
@@ -55649,31 +57145,31 @@
 set t=null
 endfunction
 function Enrage_Update takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local unit Source=(LoadUnitHandle(hash_main,(Cache),(2)))
 local integer Level=GetUnitAbilityLevel(Source,abilityid_enrage)
 local integer Bonus
 local integer LastBonus
 if GetTriggerEventId()==EVENT_UNIT_DEATH or GetUnitAbilityLevel(Source,buffid_enrage)==0 or(GetTriggerEventId()==EVENT_PLAYER_UNIT_SPELL_EFFECT and Source==GetSpellTargetUnit()and IsPurge(GetSpellAbilityId()))then
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
 call UnitRemoveAbility(Source,buffid_enrage)
 call AddDamageBonus(Source,0)
 else
-set Bonus=R2I(GetUnitState(Source,UNIT_STATE_LIFE)*(0.03+0.01*Level))
+set Bonus=R2I(GetUnitState(Source,UNIT_STATE_LIFE)*(0.04+0.01*Level))
 set LastBonus=(LoadInteger(hash_main,(Cache),(238)))
 if Bonus!=LastBonus then
 call SaveInteger(hash_main,(Cache),(238),(Bonus))
 call AddDamageBonus(Source,Bonus)
 endif
 endif
 set t=null
 set Source=null
 return false
 endfunction
 function Enrage takes nothing returns nothing
 local trigger t=CreateTrigger()
 local integer Cache=GetHandleId(t)
 call TriggerRegisterTimerEvent(t,0.25,true)
 call TriggerRegisterUnitEvent(t,GetTriggerUnit(),EVENT_UNIT_DEATH)
@@ -56182,30 +57678,31 @@
 endif
 set Hero=null
 return false
 endfunction
 function Refraction_Attacked takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local unit Hero=(LoadUnitHandle(hash_main,(GetHandleId(t)),(14)))
 local integer Cache=GetHandleId(Hero)
 local integer Remaining=(LoadInteger(hash_main,(Cache),(351)))
 if GetTriggerEventId()==EVENT_UNIT_DEATH or(GetTriggerEventId()==EVENT_PLAYER_UNIT_ATTACKED and GetAttacker()==Hero and IsUnitAlly(GetAttacker(),GetOwningPlayer(GetTriggerUnit()))==false)then
 if GetTriggerEventId()!=EVENT_UNIT_DEATH and Remaining>0 and GetEventDamage()>0 then
 call SaveInteger(hash_main,(Cache),(351),(Remaining-1))
 else
 call UnitRemoveAbility(Hero,abilityid_refractionbonusdamage)
 call UnitRemoveAbility(Hero,abilityid_refractionbonusdamagebuff)
+call UnitRemoveAbility(Hero,1110456393)
 call SaveInteger(hash_main,(Cache),(351),(0))
 endif
 endif
 set t=null
 set Hero=null
 return false
 endfunction
 function Refraction_Reset takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local unit Hero=(LoadUnitHandle(hash_main,(GetHandleId(t)),(14)))
 local integer Cache=GetHandleId(Hero)
 call SaveInteger(hash_main,(Cache),(352),(0))
 call SaveInteger(hash_main,(Cache),(351),(0))
 call UnitRemoveAbility(Hero,abilityid_refractionbonusdamage)
 call UnitRemoveAbility(Hero,abilityid_refractionbonusdamagebuff)
@@ -56213,35 +57710,35 @@
 call FlushChildHashtable(hash_main,(GetHandleId(t)))
 call CleanTrigger(t)
 set t=null
 set Hero=null
 return false
 endfunction
 function Refraction_Main takes nothing returns nothing
 local unit Hero=GetTriggerUnit()
 local integer Cache=GetHandleId(Hero)
 local trigger t=CreateTrigger()
 call TriggerRegisterTimerEvent(t,17,false)
 call TriggerAddCondition(t,Condition(function Refraction_Reset))
 call SaveUnitHandle(hash_main,(GetHandleId(t)),(14),(Hero))
 call PlaySoundAtPos(gg_snd_Vacuum,GetUnitX(Hero),GetUnitY(Hero))
 call UnitAddAbility2(Hero,abilityid_refractionpreventionbuff)
-call SaveInteger(hash_main,(Cache),(352),(GetUnitAbilityLevel(Hero,abilityid_refraction)+1))
+call SaveInteger(hash_main,(Cache),(352),(GetUnitAbilityLevel(Hero,abilityid_refraction)+2))
 call UnitAddAbility2(Hero,abilityid_refractionbonusdamage)
 call SetUnitAbilityLevel(Hero,abilityid_refractionbonusdamage,GetUnitAbilityLevel(Hero,abilityid_refraction))
 call UnitAddAbility2(Hero,abilityid_refractionbonusdamagebuff)
-call SaveInteger(hash_main,(Cache),(351),(GetUnitAbilityLevel(Hero,abilityid_refraction)+1))
+call SaveInteger(hash_main,(Cache),(351),(GetUnitAbilityLevel(Hero,abilityid_refraction)+2))
 set Hero=null
 endfunction
 function Refraction_MainLearn takes nothing returns nothing
 local unit Hero=GetTriggerUnit()
 local trigger t
 set t=CreateTrigger()
 call TriggerRegisterUnitEvent(t,Hero,EVENT_UNIT_DAMAGED)
 call TriggerRegisterUnitEvent(t,Hero,EVENT_UNIT_DEATH)
 call TriggerAddCondition(t,Condition(function Refraction_PreventDamage))
 call SaveInteger(hash_main,(GetHandleId(Hero)),(352),(0))
 set t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_ATTACKED)
 call TriggerRegisterUnitEvent(t,Hero,EVENT_UNIT_DEATH)
 call TriggerAddCondition(t,Condition(function Refraction_Attacked))
 call SaveUnitHandle(hash_main,(GetHandleId(t)),(14),(Hero))
@@ -56836,31 +58333,31 @@
 call SaveBoolean(hash_main,(Cache),(15),(Scepter))
 call SaveInteger(hash_main,(Cache),(30),(TI_AddUnit(Target)))
 call SaveReal(hash_main,(Cache),(193),((Rate)*1.0))
 call SaveInteger(hash_main,(Cache),(361),(0))
 call TriggerRegisterTimerEvent(t,20,false)
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_ATTACKED)
 call TriggerRegisterUnitEvent(t,Target,EVENT_UNIT_DAMAGED)
 call TriggerRegisterUnitEvent(t,Source,EVENT_UNIT_DEATH)
 call TriggerRegisterUnitEvent(t,Target,EVENT_UNIT_DEATH)
 call TriggerAddCondition(t,Condition(function FocusFire_Hit))
 set Source=null
 set Target=null
 set t=null
 endfunction
 function FocusFire_Main takes nothing returns boolean
-if GetSpellAbilityId()==abilityid_focusfire or GetSpellAbilityId()==abilityid_focusfireupgrade then
+if(GetSpellAbilityId()==abilityid_focusfire or GetSpellAbilityId()==abilityid_focusfireupgrade)and IsBlocked(GetSpellTargetUnit())==false then
 call FocusFire()
 endif
 return false
 endfunction
 function Register_FocusFire takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
 call TriggerAddCondition(t,Condition(function FocusFire_Main))
 set t=null
 endfunction
 function ThundergodWrath_Filter takes nothing returns boolean
 return IsUnitType(GetFilterUnit(),UNIT_TYPE_HERO)==true and GetUnitTypeId(GetFilterUnit())!=1211117642 and IsUnitEnemy(GetFilterUnit(),GetOwningPlayer(thundergodwrath_source))==true
 endfunction
 function ThundergodWrath_Enum takes nothing returns nothing
 local real x=GetUnitX(GetEnumUnit())
@@ -56912,31 +58409,31 @@
 endfunction
 constant function StaticField_WrathOfGod2Code takes nothing returns integer
 return 1093810504
 endfunction
 function StaticField_Condition takes nothing returns boolean
 return(GetSpellAbilityId()==(1093683779)or GetSpellAbilityId()==(1093677616)or GetSpellAbilityId()==(1093810503)or GetSpellAbilityId()==(1093810504))and GetUnitAbilityLevel(GetTriggerUnit(),(1093684789))>0
 endfunction
 function StaticField_ShockThem takes nothing returns nothing
 local real CurrentLife=GetUnitState(GetEnumUnit(),UNIT_STATE_LIFE)
 local real LifeFactor=0.03+0.02*GetUnitAbilityLevel(GetTriggerUnit(),(1093684789))
 local real Damage=CurrentLife*LifeFactor
 call UnitDamageTarget(GetTriggerUnit(),GetEnumUnit(),Damage,true,true,ATTACK_TYPE_NORMAL,DAMAGE_TYPE_MAGIC,WEAPON_TYPE_WHOKNOWS)
 call DestroyEffect(AddSpecialEffectTarget("Abilities\\Spells\\Other\\ForkedLightning\\ForkedLightningTarget.mdl",GetEnumUnit(),"overhead"))
 endfunction
 function StaticField_WhichToShock takes nothing returns boolean
-return IsUnitType(GetFilterUnit(),UNIT_TYPE_HERO)==true and IsUnitEnemy(GetFilterUnit(),GetOwningPlayer(GetTriggerUnit()))and IsUnitVisible(GetFilterUnit(),GetOwningPlayer(GetTriggerUnit()))
+return GetUnitAbilityLevel(GetFilterUnit(),1093678162)==0 and IsUnitEnemy(GetFilterUnit(),GetOwningPlayer(GetTriggerUnit()))and IsUnitType(GetFilterUnit(),UNIT_TYPE_STRUCTURE)==false and IsDead(GetFilterUnit())==false
 endfunction
 function StaticField_Action takes nothing returns nothing
 local location PointVar=GetUnitLoc(GetTriggerUnit())
 local group GroupVar=GetAvailableGroup()
 local real x=GetLocationX(PointVar)
 local real y=GetLocationY(PointVar)
 local boolexpr CondVar=Condition(function StaticField_WhichToShock)
 call GroupEnumUnitsInRange(GroupVar,x,y,1025,CondVar)
 call ForGroup(GroupVar,function StaticField_ShockThem)
 call RemoveLocation(PointVar)
 call KillGroup(GroupVar)
 endfunction
 function RegisterStaticField takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
@@ -57088,34 +58585,38 @@
 set Source=(LoadUnitHandle(hash_main,(Cache),(2)))
 set Level=(LoadInteger(hash_main,(Cache),(5)))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(32))))
 call SaveBoolean(hash_main,(GetHandleId(Target)),(363),(false))
 call TI_ClearUnit(TargetIndex)
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
 call PreventDamage(GetTriggerUnit(),Life)
 call AphoticShield_Explode(Source,GetTriggerUnit(),Level)
 elseif GetEventDamage()>0 then
 call PreventDamage(GetTriggerUnit(),GetEventDamage())
 call SaveReal(hash_main,(Cache),(362),((Life-GetEventDamage())*1.0))
 endif
 endif
 else
+set Source=(LoadUnitHandle(hash_main,(Cache),(2)))
+set Level=(LoadInteger(hash_main,(Cache),(5)))
 call SaveBoolean(hash_main,(GetHandleId(Target)),(363),(false))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(32))))
+call TI_ClearUnit(TargetIndex)
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
+call AphoticShield_Explode(Source,Target,Level)
 endif
 set t=null
 return false
 endfunction
 function AphoticShield takes nothing returns nothing
 local trigger t=CreateTrigger()
 local integer Cache=GetHandleId(t)
 local unit Target=GetSpellTargetUnit()
 local unit Source=GetTriggerUnit()
 local integer Level=GetUnitAbilityLevel(Source,abilityid_aphoticshield)
 local integer Cache2=GetHandleId(Target)
 local real Life
 if(LoadBoolean(hash_main,(Cache2),(363)))==true then
 call TriggerEvaluate((LoadTriggerHandle(hash_main,(Cache2),(364))))
 endif
@@ -59193,60 +60849,62 @@
 set Target=null
 return false
 endfunction
 function AstralImprisonment_Finish takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local unit Source=(LoadUnitHandle(hash_main,(Cache),(2)))
 local unit Target=(LoadUnitHandle(hash_main,(Cache),(17)))
 call SaveInteger(hash_main,(GetHandleId((Target))),((4303)),(HERO_STATE_OFF))
 call DestroyEffect((LoadEffectHandle(hash_main,(Cache),(32))))
 call SetUnitInvulnerable(Target,false)
 call PauseUnit(Target,false)
 call ShowUnit(Target,true)
 call ClearSelectionForPlayer(GetOwningPlayer(Target))
 call SelectUnitAddForPlayer(Target,GetOwningPlayer(Target))
+call RemoveSavedHandle(hash_main,(GetHandleId(Source)),(748))
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
 set t=null
 set Source=null
 set Target=null
 return false
 endfunction
 function AstralImprisonment takes nothing returns nothing
 local unit Source=GetTriggerUnit()
 local unit Target=GetSpellTargetUnit()
 local integer Level=GetUnitAbilityLevel(Source,abilityid_astralimprisonment)
 local integer StolenInt=IMin(GetHeroInt(Target,false),Level*2)
 local trigger t
 local integer Cache
 local unit Caster=CreateUnit(GetOwningPlayer(Target),1865429305,GetUnitX(Target),GetUnitY(Target),0)
 call UnitApplyTimedLife(Caster,1112820806,Level+1)
 if(IsUnitEnemy(Target,GetOwningPlayer(Source)))==false or IsBlocked(GetSpellTargetUnit())==false then
 call SaveInteger(hash_main,(GetHandleId((Target))),((4303)),(HERO_STATE_ON))
 set t=CreateTrigger()
 set Cache=GetHandleId(t)
 call SelectUnitRemoveForPlayer(Target,GetOwningPlayer(Target))
 call SetUnitInvulnerable(Target,true)
 call PauseUnit(Target,true)
 call ShowUnit(Target,false)
 call SaveEffectHandle(hash_main,(Cache),(32),(AddSpecialEffect("Abilities\\Spells\\Demon\\DarkConversion\\ZombifyTarget.mdl",GetUnitX(Target),GetUnitY(Target))))
 call SaveUnitHandle(hash_main,(Cache),(2),(Source))
 call SaveUnitHandle(hash_main,(Cache),(17),(Target))
 call TriggerRegisterTimerEvent(t,Level,false)
 call TriggerAddCondition(t,Condition(function AstralImprisonment_Finish))
 if(IsUnitEnemy(Target,GetOwningPlayer(Source)))==true then
+call SaveUnitHandle(hash_main,(GetHandleId(Source)),(748),(Target))
 call SetHeroInt(Target,GetHeroInt(Target,false)-StolenInt,true)
 call SetHeroInt(Source,GetHeroInt(Source,false)+StolenInt,true)
 set t=CreateTrigger()
 set Cache=GetHandleId(t)
 call SaveUnitHandle(hash_main,(Cache),(2),(Source))
 call SaveUnitHandle(hash_main,(Cache),(17),(Target))
 call SaveInteger(hash_main,(Cache),(262),(StolenInt))
 call TriggerRegisterTimerEvent(t,60,false)
 call TriggerRegisterUnitEvent(t,Source,EVENT_UNIT_DEATH)
 call TriggerRegisterUnitEvent(t,Target,EVENT_UNIT_DEATH)
 call TriggerAddCondition(t,Condition(function AstralImprisonment_RestoreInt))
 endif
 endif
 set Source=null
 set Target=null
@@ -59308,71 +60966,79 @@
 function SanitysEclipseWhichUnits takes nothing returns boolean
 if(IsUnitType(GetFilterUnit(),UNIT_TYPE_HERO)==true)then
 if(IsUnitEnemy(GetFilterUnit(),GetOwningPlayer(GetTriggerUnit())))then
 return true
 endif
 endif
 return false
 endfunction
 function Trig_Sanitys_Eclipse_Actions takes nothing returns nothing
 local unit UnitVar=GetTriggerUnit()
 local integer level=GetUnitAbilityLevel(UnitVar,1093685067)
 local real x=GetLocationX(GetSpellTargetLoc())
 local real y=GetLocationY(GetSpellTargetLoc())
 local group GroupVar=GetAvailableGroup()
 local integer offset=0
+local unit AstralTarget=(LoadUnitHandle(hash_main,(GetHandleId(UnitVar)),(748)))
 if level==0 then
 set level=GetUnitAbilityLevel(UnitVar,abilityid_sanityseclipse_upgrade)
 endif
 call DestroyEffect(AddSpecialEffect("Abilities\\Spells\\Human\\MarkOfChaos\\MarkOfChaosTarget.mdl",x,y))
 loop
 exitwhen offset>=360
 call DestroyEffect(AddSpecialEffect("Abilities\\Spells\\Human\\MarkOfChaos\\MarkOfChaosTarget.mdl",x+(200+level*50)*Cos(offset*bj_DEGTORAD),y+(200+level*50)*Sin(offset*bj_DEGTORAD)))
 set offset=offset+45
 endloop
 call GroupEnumUnitsInRange(GroupVar,x,y,300+level*100,Condition(function SanitysEclipseWhichUnits))
+if AstralTarget!=null and DistanceBetweenXY(x,y,GetUnitX(AstralTarget),GetUnitY(AstralTarget))<(300+level*100)then
+call SetUnitInvulnerable(AstralTarget,false)
+call GroupAddUnit(GroupVar,AstralTarget)
+endif
 call ForGroup(GroupVar,function SanitysEclipseDamage)
+if AstralTarget!=null and IsDead(AstralTarget)==false then
+call SetUnitInvulnerable(AstralTarget,true)
+endif
 call KillGroup(GroupVar)
 endfunction
 function RegisterSanitys_Eclipse takes nothing returns nothing
 local trigger t=CreateTrigger()
 call TriggerRegisterAnyUnitEvent(t,EVENT_PLAYER_UNIT_SPELL_EFFECT)
 call TriggerAddCondition(t,Condition(function Trig_Sanitys_Eclipse_Conditions))
 call TriggerAddAction(t,function Trig_Sanitys_Eclipse_Actions)
 endfunction
 function IonShell_Damage takes nothing returns nothing
 
@@ -62326,31 +63993,31 @@
 call IssueImmediateOrder(Caster,"thunderclap")
 call SelectUnitForPlayerSingle(Source,GetOwningPlayer(Source))
 call FlushChildHashtable(hash_main,(Cache))
 call CleanTrigger(t)
 set t=null
 set Source=null
 set Caster=null
 return false
 endfunction
 function Reincarnation_Dead takes nothing returns boolean
 local trigger t=GetTriggeringTrigger()
 local integer Cache=GetHandleId(t)
 local unit Source=(LoadUnitHandle(hash_main,(Cache),(2)))
 local integer Level=GetUnitAbilityLevel(Source,abilityid_reincarnation)
 if TimerGetRemaining(CTimer[GetPlayerId(GetOwningPlayer(Source))])==0 and GetUnitState(Source,UNIT_STATE_MANA)>=140 then
-call TimerStart(CTimer[GetPlayerId(GetOwningPlayer(Source))],380-80*Level,false,null)
+call TimerStart(CTimer[GetPlayerId(GetOwningPlayer(Source))],360-100*Level,false,null)
 set t=CreateTrigger()
 set Cache=GetHandleId(t)
 call TriggerRegisterTimerEvent(t,3+0.01,false)
 call TriggerAddCondition(t,Condition(function Reincarnation_Control))
 call SaveUnitHandle(hash_main,(Cache),(2),(Source))
 endif
 set t=null
 set Source=null
 return false
 endfunction
 function Reincarnation_Learned takes nothing returns nothing
 local trigger t=CreateTrigger()
 local integer Cache=GetHandleId(t)
 local unit Source=GetTriggerUnit()
 call TriggerRegisterDeathEvent(t,Source)
@@ -78839,33 +84851,35 @@
 call InitTrig_Misc()
 call InitTrig_Map_Info()
 call InitTrig_Leave()
 call InitTrig_Remove_On_Respawn()
 call InitTrig_HeroRevive()
 call InitTrig_HeroEnters()
 call InitTrig_OneHero()
 call InitTrig_Money()
 call InitTrig_Time()
 call InitTrig_Cooldown_Tracker()
 call InitTrig_Reliable_Gold()
 call InitTrig_Ban_Dev_Hero()
 call InitTrig_Track_Dying_Creeps()
 call InitTrig_CustomHeroRevive()
 call InitTrig_Deathtrigger()
call InitTrig_Image_Enters()
 call InitTrig_First_Blood()
 call InitTrig_RoshanDies()
 call InitTrig_AntiDeny()
+call InitTrig_Chicken_Dies()
 call InitTrig_Item_Definitions()
 call InitTrig_Powerup_Definitions()
 call Initialize_Recipes()
 call AghanimsScepter_Init()
 call InitTrig_Goblin_Definitions()
 call InitTrig_Unit_Item_Conversion()
 call InitTrig_Manipulate_Item()
 call InitTrig_Hero_Morphing_Restrictions()
 call InitTrig_Item_Traffic()
 call InitTrig_Item_Cooldown()
 call InitTrig_Disassemble()
 call InitTrig_NoChickenSale()
 call InitTrig_Bloodstone()
 call InitTrig_Bloodstone_Heal()
 call InitTrig_NoRoshanWard()
